### PR TITLE
Add nsfallback feature for browser request

### LINKED ISSUE #2261: Feature Request: Support for "reverse default namespace linking"
Namespace default linking (https://www.dokuwiki.org/namespaces#namespace_default_linking) implements a kind of "fallback" in case someone addresses `foo:bar:` by searching `foo:bar:$conf['start']` -> `foo:bar:bar` -> `foo:bar` and using the first found.

However in daily usage, the other way round  would be more useful: When referencing `foo:bar` DW could search `foo:bar` -> `foo:bar:$conf['start']` -> `foo:bar:bar` and display the first found.  This would result in an user experience like we are used from internet browsing when http://somewhere.else/page is - if appropriate - "automatically" expanded to http://somewhere.else/page/index.html.  It would also save users from always having to type the trailing / when they want to address the namespace.

I know that this could also be achieved by rewrite rules directly in the webserver, but IMHO it would be nicer if DW could/would handle this itself.

### PR DESCRIPTION
**Pull request by phy25**
nsfallback is a fallback ns search when requesting:
foo:bar -> foo:bar:$conf['start'] -> foo:bar:bar

nsfallback is similar to autoplural, but autoplural is only effective during internal link parsing (not effective during browser request i.e. getID()). Not sure why it works like that, but because it's not rewriting getID you can try creating page by visiting the ID directly - however it is not working by searching for the not existed page or linking it, which is the recommended way in https://www.dokuwiki.org/page#create_a_page.

I decided to implement nsfallback to apply only during browser request and $ACT=show, not during internal link parsing, because this is most useful when users visit /foo/bar looking for /foo/bar/(start) - this is less intended for people writing the link on the wiki. By not rewriting the internal link, user can still create pages by linking it or searching.

This option is off by default. Tests for getid and resolve_pageid have been added to define the behavior.

This closes #2261. *Tests will go on once #2850 is merged, and I can do a rebase after that.*

### CODE REVIEW

File: _test/tests/inc/pageutils_getid.test.php [hunk 1]
```diff
@@ -14,7 +14,7 @@ function test_zero_id(){
         $_SERVER['REQUEST_URI'] = '/doku.php?id=0&do=edit';
         $_REQUEST['id'] = '0';
 
-        $this->assertSame('0', getID('id'));
+        $this->assertSame(getID('id'), '0');
```
Comment (splitbrain) [inline _test/tests/inc/pageutils_getid.test.php:17]: This one was the right way around already ;-)

File: _test/tests/inc/pageutils_getid.test.php [hunk 2]
```diff
@@ -114,8 +114,109 @@ function test5() {
         $_SERVER['PATH_TRANSLATED'] = '/var/www/index.html';
         $_SERVER['PHP_SELF'] = '/dokuwiki/doku.php/';
 
-        $this->assertEquals(getID(), cleanID($conf['start']));
+        $this->assertEquals(cleanID($conf['start']), getID());
     }
 
+    function _test_default_ns($path, $expected){
```
Comment (splitbrain) [inline _test/tests/inc/pageutils_getid.test.php:120]: Why the `_` prefix?

Comment (splitbrain) [reply _test/tests/inc/pageutils_getid.test.php:120]: And there's no dataProvider set?

**Comment by phy25** [comment]
So this can be separated into several parts with the potential name `RewriteRule` and `RewriteRouter`:

- read rewritten URL from the request
- namespace to start page
- not-found page with the same name as namespace to namespace start page (nsfallback)

I am thinking of something like `ParserMode` that has a sort order, but I am not sure if it is too complicated. Or I would create an event here and let those rules be hooks and use sequence numbers (then I have to figure out how to autoload those classes without having them in a plugin). What do you think @splitbrain?

**Pull request closed without merging**